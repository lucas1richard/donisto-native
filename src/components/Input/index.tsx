import React from 'react';
import { View, TextInput } from 'react-native';
import { IFieldPrimitiveProps } from 'components/FieldPrimitive';
import showToast from 'components/Toast';
import theme from 'theme/iftheme';
import { WrappedFieldProps } from 'redux-form';
import { BlurView } from 'expo';

export type TFieldLabels = 'fieldLabel'|'fieldLabelerror'|'fieldLabelactive'|'fieldLabelactive';
export type TInputLabels = 'inputStyle'|'inputStyleerror'|'inputStyleactive'|'inputStyleactive';
export type TTextLabels = 'textInputWrapper'|'textInputWrappererror'|'textInputWrapperactive'|'textInputWrapperactive';

export interface IStyleKeys {
  fieldLabelKey: TFieldLabels;
  inputStyleKey: TInputLabels;
  textInputWrapperKey: TTextLabels;
}

export interface IFieldPrimitiveProps {
  component: any;
  noToast?: boolean;
  validateAtLength?: number;
  input?: any;
  meta?: any;
  name: string;
  keyboardType?: string;
  label?: string;
}

class Input extends React.Component<WrappedFieldProps & IFieldPrimitiveProps> {
  onBlur = () => {
    const { input, meta, noToast } = this.props;
    if (meta.error && !noToast) {
      const { error } = meta;
      const text = Array.isArray(error) ? error[0] : error;
      showToast({ text: `${text}` });
    }
    if (input && input.onBlur) {
      input.onBlur(input.value);
    }
  }

  onFocus = (ev: any) => {
    this.props.input.onFocus(ev);
  }

  styleKeys(): IStyleKeys {
    const {
      input,
      meta,
      validateAtLength
    } = this.props;

    let readyToShow;

    if (validateAtLength) {
      readyToShow = input.value.length === validateAtLength;
    } else {
      readyToShow = meta.touched || (input.value !== '' && meta.pristine);
    }

    const hasError = readyToShow && meta.invalid;
    const isValid = readyToShow && meta.valid;

    let status: ''|'error'|'active'|'valid' = '';
    if (hasError) {
      status = 'error';
    } else if (meta.active) {
      status = 'active';
    } else if (isValid) {
      status = 'valid';
    }

    const fieldLabelKey = `fieldLabel${status}` as TFieldLabels;
    const inputStyleKey = `inputStyle${status}` as TInputLabels;
    const textInputWrapperKey = `textInputWrapper${status}` as TTextLabels;

    /**
     * These keys are the same as keys in styles.js
     * files for components which extend FieldPrimitive
     */
    return {
      fieldLabelKey,
      inputStyleKey,
      textInputWrapperKey
    };
  }

  render() {
    const {
      noMargin,
      style,
      input,
      innerRef,
      validateAtLength,
      specialError,
      showFloatingLabel,
      multiline,
      meta,
      icon,
      ...field
    } = this.props;
    // const {
    //   fieldLabelKey
    // } = this.styleKeys();

    const hasIcon = !!icon;

    return (
        <View>
          <BlurView tint="dark" intensity={80} style={{ borderRadius: theme.fontSize * 0.6 }}>
            <View style={{ flexDirection: 'row', alignItems: 'center' }}>
              {hasIcon && (
                <View style={{ paddingLeft: theme.fontSize }}>
                  {icon}
                </View>
              )}
              <TextInput
                {...input}
                {...field}
                allowFontScaling={false}
                ref={innerRef}
                label={field.label}
                placeholder={field.label}
                placeholderTextColor="#bbb"
                value={input.value}
                onChangeText={input.onChange}
                tintColor={theme.green}
                style={{ color: '#fff', fontSize: theme.fontSize + 4, padding: theme.fontSize }}
                onBlur={this.onBlur}
                onFocus={this.onFocus}
                error={meta.submitFailed ? meta.error : ''}
                returnKeyType={field.returnKeyType || 'next'}
              />
            </View>
          </BlurView>
        </View>
    );
  }
}

export default Input;
