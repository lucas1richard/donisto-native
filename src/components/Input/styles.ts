import { StyleSheet, ViewStyle, TextStyle } from 'react-native';
import theme from 'theme/iftheme';

export const style = {
  wrapper: {
    backgroundColor: '#fff',
    width: '100%'
  },
  errortrue: error(true),
  errorfalse: error(false),
  fieldLabel: fieldLabel(),
  fieldLabelerror: fieldLabel({ color: theme.red }),
  fieldLabelvalid: fieldLabel({ color: theme.green }),
  fieldLabelactive: fieldLabel({ color: theme.darkBlue }),
  textInputWrapper: textInputWrapper({ borderColor: theme.pristineInputGray }),
  textInputWrappererror: textInputWrapper({ borderColor: theme.red }),
  textInputWrappervalid: textInputWrapper({ borderColor: theme.green }),
  textInputWrapperactive: textInputWrapper({ borderColor: theme.darkBlue }),
  inputStyle: inputStyle(),
  inputStyleerror: inputStyle({ color: theme.red }),
  inputStylevalid: inputStyle({ color: theme.green }),
  inputStyleactive: inputStyle({ color: theme.darkBlue })
};

export default StyleSheet.create(style);

function inputStyle(addStyle = {}): TextStyle {
  return {
    width: '100%',
    color: theme.darkGray,
    position: 'relative',
    padding: 12,
    fontFamily: 'proxima-soft-regular',
    fontWeight: '300',
    height: 120,
    fontSize: 18,
    marginTop: -35,
    ...addStyle
  };
}

function textInputWrapper(addStyle = {}): ViewStyle {
  return {
    borderRadius: 4,
    borderWidth: 1,
    borderStyle: 'solid',
    borderColor: theme.darkGray,
    backgroundColor: '#fff',
    width: '100%',
    height: 50,
    overflow: 'hidden',
    ...addStyle
  };
}

/** The input label */
function fieldLabel(addStyle = {}): TextStyle {
  return ({
    color: theme.darkGray,
    backgroundColor: 'rgba(0,0,0,0)',
    marginBottom: 3,
    fontWeight: '300',
    fontSize: 12,
    ...addStyle
  });
}

/**
 * Error text style
 */
function error(err: boolean): TextStyle {
  return {
    color: theme.red,
    fontSize: 12,
    height: 16,
    opacity: err ? 1 : 0
  };
}
