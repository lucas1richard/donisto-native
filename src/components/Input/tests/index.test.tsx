import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Input, { TInputProps } from '..';
import { BlurViewProps } from 'expo';

const defaultProps = {
  meta: {
    autofilled: false,
    asyncValidating: false,
    dirty: false,
    dispatch: jest.fn(),
    form: 'mockForm',
    initial: '',
    invalid: false,
    pristine: true,
    submitting: false,
    submitFailed: false,
    touched: false,
    valid: true,
    visited: false
  },
  input: {
    value: 'mockValue',
    onBlur: jest.fn(),
    onChange: jest.fn(),
    name: 'mockName',
    onDragStart: jest.fn(),
    onDrop: jest.fn(),
    onFocus: jest.fn()
  },
};

describe('<Input />', () => {
  const renderComponent = (props: TInputProps) => shallow(<Input {...props} />);
  it('renders', () => {
    expect(toJson(renderComponent({ ...defaultProps }))).toMatchSnapshot();
  });
  it('renders multiline', () => {
    expect(renderComponent({
      ...defaultProps,
      multiline: true
    })).toMatchSnapshot();
  });
  it('focuses', () => {
    const onFocus = jest.fn();
    const instance = renderComponent({
      ...defaultProps,
      multiline: true,
      input: {
        ...defaultProps.input,
        onFocus
      }
    }).instance() as Input;
    instance.onFocus();
    expect(onFocus).toHaveBeenCalled();
  });
  describe('onBlur instance method', () => {
    it('blurs', () => {
      const onBlur = jest.fn();
      const instance = renderComponent({
        ...defaultProps,
        multiline: true,
        input: {
          ...defaultProps.input,
          onBlur
        }
      }).instance() as Input;
      instance.onBlur();
      expect(onBlur).toHaveBeenCalled();
    });
  });
  describe('blurViewProps', () => {
    it('renders with blurViewProps', () => {
      const blurViewProps: BlurViewProps = {
        tint: 'light',
        intensity: 80
      };
      expect(toJson(renderComponent({
        ...defaultProps,
        blurViewProps
      }))).toMatchSnapshot();
    });
  });
});