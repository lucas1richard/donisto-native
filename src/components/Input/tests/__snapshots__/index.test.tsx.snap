// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> blurViewProps renders with blurViewProps 1`] = `
<Component>
  <BlurView
    intensity={80}
    style={
      Array [
        Object {
          "borderRadius": 12,
        },
        undefined,
      ]
    }
    tint="light"
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={false}
        error=""
        name="mockName"
        numberOfLines={1}
        onBlur={[Function]}
        onChange={[MockFunction]}
        onChangeText={[MockFunction]}
        onDragStart={[MockFunction]}
        onDrop={[MockFunction]}
        onFocus={[Function]}
        placeholderTextColor="#333"
        returnKeyType="next"
        style={
          Object {
            "color": "#000",
            "flex": 14,
            "fontSize": 18,
            "minHeight": undefined,
            "padding": 20,
          }
        }
        textAlignVertical="center"
        underlineColorAndroid="transparent"
        value="mockValue"
      />
    </Component>
  </BlurView>
</Component>
`;

exports[`<Input /> renders 1`] = `
<Component>
  <BlurView
    intensity={80}
    style={
      Array [
        Object {
          "borderRadius": 12,
        },
        undefined,
      ]
    }
    tint="default"
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={false}
        error=""
        name="mockName"
        numberOfLines={1}
        onBlur={[Function]}
        onChange={[MockFunction]}
        onChangeText={[MockFunction]}
        onDragStart={[MockFunction]}
        onDrop={[MockFunction]}
        onFocus={[Function]}
        placeholderTextColor="#bbb"
        returnKeyType="next"
        style={
          Object {
            "color": "#fff",
            "flex": 14,
            "fontSize": 18,
            "minHeight": undefined,
            "padding": 20,
          }
        }
        textAlignVertical="center"
        underlineColorAndroid="transparent"
        value="mockValue"
      />
    </Component>
  </BlurView>
</Component>
`;

exports[`<Input /> renders multiline 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    input={
      Object {
        "name": "mockName",
        "onBlur": [MockFunction],
        "onChange": [MockFunction],
        "onDragStart": [MockFunction],
        "onDrop": [MockFunction],
        "onFocus": [MockFunction],
        "value": "mockValue",
      }
    }
    meta={
      Object {
        "asyncValidating": false,
        "autofilled": false,
        "dirty": false,
        "dispatch": [MockFunction],
        "form": "mockForm",
        "initial": "",
        "invalid": false,
        "pristine": true,
        "submitFailed": false,
        "submitting": false,
        "touched": false,
        "valid": true,
        "visited": false,
      }
    }
    multiline={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <BlurView
        intensity={80}
        style={
          Array [
            Object {
              "borderRadius": 12,
            },
            undefined,
          ]
        }
        tint="default"
      >
        <Component
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            allowFontScaling={false}
            error=""
            name="mockName"
            numberOfLines={1}
            onBlur={[Function]}
            onChange={[MockFunction]}
            onChangeText={[MockFunction]}
            onDragStart={[MockFunction]}
            onDrop={[MockFunction]}
            onFocus={[Function]}
            placeholderTextColor="#bbb"
            returnKeyType="next"
            style={
              Object {
                "color": "#fff",
                "flex": 14,
                "fontSize": 18,
                "minHeight": undefined,
                "padding": 20,
              }
            }
            textAlignVertical="center"
            underlineColorAndroid="transparent"
            value="mockValue"
          />
        </Component>
      </BlurView>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Component
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            allowFontScaling={false}
            error=""
            name="mockName"
            numberOfLines={1}
            onBlur={[Function]}
            onChange={[MockFunction]}
            onChangeText={[MockFunction]}
            onDragStart={[MockFunction]}
            onDrop={[MockFunction]}
            onFocus={[Function]}
            placeholderTextColor="#bbb"
            returnKeyType="next"
            style={
              Object {
                "color": "#fff",
                "flex": 14,
                "fontSize": 18,
                "minHeight": undefined,
                "padding": 20,
              }
            }
            textAlignVertical="center"
            underlineColorAndroid="transparent"
            value="mockValue"
          />
        </Component>,
        "intensity": 80,
        "style": Array [
          Object {
            "borderRadius": 12,
          },
          undefined,
        ],
        "tint": "default",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            false,
            <TextInput
              allowFontScaling={false}
              error=""
              name="mockName"
              numberOfLines={1}
              onBlur={[Function]}
              onChange={[MockFunction]}
              onChangeText={[MockFunction]}
              onDragStart={[MockFunction]}
              onDrop={[MockFunction]}
              onFocus={[Function]}
              placeholderTextColor="#bbb"
              returnKeyType="next"
              style={
                Object {
                  "color": "#fff",
                  "flex": 14,
                  "fontSize": 18,
                  "minHeight": undefined,
                  "padding": 20,
                }
              }
              textAlignVertical="center"
              underlineColorAndroid="transparent"
              value="mockValue"
            />,
          ],
          "style": Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
        },
        "ref": null,
        "rendered": Array [
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "allowFontScaling": false,
              "error": "",
              "label": undefined,
              "multiline": undefined,
              "name": "mockName",
              "numberOfLines": 1,
              "onBlur": [Function],
              "onChange": [MockFunction],
              "onChangeText": [MockFunction],
              "onDragStart": [MockFunction],
              "onDrop": [MockFunction],
              "onFocus": [Function],
              "placeholder": undefined,
              "placeholderTextColor": "#bbb",
              "returnKeyType": "next",
              "style": Object {
                "color": "#fff",
                "flex": 14,
                "fontSize": 18,
                "minHeight": undefined,
                "padding": 20,
              },
              "textAlignVertical": "center",
              "underlineColorAndroid": "transparent",
              "value": "mockValue",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <BlurView
          intensity={80}
          style={
            Array [
              Object {
                "borderRadius": 12,
              },
              undefined,
            ]
          }
          tint="default"
        >
          <Component
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              allowFontScaling={false}
              error=""
              name="mockName"
              numberOfLines={1}
              onBlur={[Function]}
              onChange={[MockFunction]}
              onChangeText={[MockFunction]}
              onDragStart={[MockFunction]}
              onDrop={[MockFunction]}
              onFocus={[Function]}
              placeholderTextColor="#bbb"
              returnKeyType="next"
              style={
                Object {
                  "color": "#fff",
                  "flex": 14,
                  "fontSize": 18,
                  "minHeight": undefined,
                  "padding": 20,
                }
              }
              textAlignVertical="center"
              underlineColorAndroid="transparent"
              value="mockValue"
            />
          </Component>
        </BlurView>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Component
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              allowFontScaling={false}
              error=""
              name="mockName"
              numberOfLines={1}
              onBlur={[Function]}
              onChange={[MockFunction]}
              onChangeText={[MockFunction]}
              onDragStart={[MockFunction]}
              onDrop={[MockFunction]}
              onFocus={[Function]}
              placeholderTextColor="#bbb"
              returnKeyType="next"
              style={
                Object {
                  "color": "#fff",
                  "flex": 14,
                  "fontSize": 18,
                  "minHeight": undefined,
                  "padding": 20,
                }
              }
              textAlignVertical="center"
              underlineColorAndroid="transparent"
              value="mockValue"
            />
          </Component>,
          "intensity": 80,
          "style": Array [
            Object {
              "borderRadius": 12,
            },
            undefined,
          ],
          "tint": "default",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              false,
              <TextInput
                allowFontScaling={false}
                error=""
                name="mockName"
                numberOfLines={1}
                onBlur={[Function]}
                onChange={[MockFunction]}
                onChangeText={[MockFunction]}
                onDragStart={[MockFunction]}
                onDrop={[MockFunction]}
                onFocus={[Function]}
                placeholderTextColor="#bbb"
                returnKeyType="next"
                style={
                  Object {
                    "color": "#fff",
                    "flex": 14,
                    "fontSize": 18,
                    "minHeight": undefined,
                    "padding": 20,
                  }
                }
                textAlignVertical="center"
                underlineColorAndroid="transparent"
                value="mockValue"
              />,
            ],
            "style": Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
          },
          "ref": null,
          "rendered": Array [
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "allowFontScaling": false,
                "error": "",
                "label": undefined,
                "multiline": undefined,
                "name": "mockName",
                "numberOfLines": 1,
                "onBlur": [Function],
                "onChange": [MockFunction],
                "onChangeText": [MockFunction],
                "onDragStart": [MockFunction],
                "onDrop": [MockFunction],
                "onFocus": [Function],
                "placeholder": undefined,
                "placeholderTextColor": "#bbb",
                "returnKeyType": "next",
                "style": Object {
                  "color": "#fff",
                  "flex": 14,
                  "fontSize": 18,
                  "minHeight": undefined,
                  "padding": 20,
                },
                "textAlignVertical": "center",
                "underlineColorAndroid": "transparent",
                "value": "mockValue",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
